pkg/gpu/context.go

```diff
@@ -209,13 +218,67 @@ func (ctx *systemContext) cleanupOldEntries() {
 	}
 }
 
+// filterDevicesForContainer filters the available GPU devices for the given
+// container. If the ID is not empty, we check the assignment of GPU resources
+// to the container and return only the devices that are available to the
+// container.
+func (ctx *systemContext) filterDevicesForContainer(devices []nvml.Device, containerID string) []nvml.Device {
+	if containerID == "" {
+		// If the process is not running in a container, we assume all devices are available.
+		return devices
+	}
+
+	container, err := ctx.workloadmeta.GetContainer(containerID)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.Warnf("Error getting container %s: %s", containerID, err)
+		}
+
+		return devices
+	}
+
+	var filteredDevices []nvml.Device
+	for _, resource := range container.AllocatedResources {
+		// Only consider NVIDIA GPUs
+		if resource.Name != nvidiaResourceName {
+			continue
+		}
+
+		for _, device := range devices {
+			uuid, ret := device.GetUUID()
+			if ret != nvml.SUCCESS {
+				log.Warnf("Error getting GPU UUID for device %s: %s", device, nvml.ErrorString(ret))
+				continue
+			}
+
+			if resource.ID == uuid {
+				filteredDevices = append(filteredDevices, device)
+			}
```

Comment:
do you need to continue the iteration of the devices (inner for loop) for the given resource after the match?

Reply:
Changed!